MTKA-1370: Save post titles to posts table

mindctrl opened the pull request:
## Description
This saves title field data to `wp_posts.post_title` instead of `wp_postmeta` like it currently does. As a result, the title filtering logic has been removed.

When fetching a title field's value, if it's in postmeta it is migrated to the posts table and removed from postmeta.

Title data is still available in WPGraphQL as the field name as well as `title`.

Related, the post name/slug is generated from the title when available instead of using the post ID or "entry{id}".

Closes #388 
Closes #280 

https://wpengine.atlassian.net/browse/MTKA-1370

## Testing
Integration tests have been updated to test desired behavior. Please double check the behavior that you see when testing manually.

mindctrl committed:
wip: stub out get/save field function. [skip ci]

matthewguywright committed:
Update class-publisher-form-editing-experience.php

matthewguywright committed:
Update class-publisher-form-editing-experience.php

matthewguywright committed:
Update class-publisher-form-editing-experience.php

mindctrl committed:
chore: fix misspelled word in mock data.

mindctrl committed:
refactor: remove post title filtering logic.

This will no longer be necessary once titles are saved in `wp_posts.post_title`.

chore: remove post title filter tests.

mindctrl committed:
refactor: simplify field data fetching logic passed to `wp_localize_script`.

mindctrl committed:
refactor: add additional field types to `get_field_value()`.

When fetching post title fields, if the data is saved in `wp_postmeta`
it will be migrated to the `wp_posts` table.

mindctrl committed:
test: remove check that `filter_post_titles` callback is attached to action.

This is no longer needed since we store titles in wp_posts where they belong.

mindctrl committed:
test: check that post title value matches mock data value.

mindctrl committed:
test: check that post slug matches mock data slug.

mindctrl committed:
test: check that post ID in REST API response matches post ID in mock data.

Previously this checked that the slug was the ID. Now the slug is based on the post title.

mindctrl committed:
test: check that post ID in REST API response matches post ID in mock data.

Previously this checked that the slug was the ID. Now the slug is based on the post title.

mindctrl committed:
test: check that title in GraphQL response matches title on mock data.

Previously this test ignored the mock title and checked postmeta for the value.

mindctrl committed:
refactor: return post title from posts table in GraphQL responses.

Also migrates any existing title data from wp_postmeta to wp_posts.

mindctrl committed:
test: change expected fallback title value when a title field does not exist.

Previously it checked for "No Title", and now it checks that it starts with "entry"
to match the fallback title generation logic.

mindctrl committed:
refactor: update field save/delete logic to include other field types.

mindctrl committed:
refactor: adjust logic and return values for `save_relationship_field()` method.

Previously it returned void, which made it hard to check for errors, and doesn't match
the upstream library return value. Now it returns a boolean.

mindctrl committed:
refactor: adjust `save_post` logic.

Simplifies logic and uses helper methods for get/save/delete actions.

mindctrl committed:
chore: coding standards

mindctrl committed:
chore: add changelog entry

mindctrl committed:
chore: fix return value when no value exists in db

nickcernis reviewed (commented):
The title functionality itself is working well for me!

<img width="411" alt="Screenshot 2022-02-28 at 13 52 10" src="https://user-images.githubusercontent.com/647669/155986997-c5188768-75c5-4739-861f-ddf649c28b7e.png">

For the migration logic, one side-effect is that all post titles revert to their original slugs until you visit the post (or cause the title to be fetched in some way).

For example, here's my original Actors admin page index (main branch, actors created with a title text field before checking out this branch):

<img width="471" alt="Screenshot 2022-02-28 at 13 54 15" src="https://user-images.githubusercontent.com/647669/155987205-60c6a48d-e8ac-45fd-9a0b-6acbd0c20d6d.png">

And here it is as soon as I switch to this PR's branch:

<img width="443" alt="Screenshot 2022-02-28 at 13 54 03" src="https://user-images.githubusercontent.com/647669/155987244-594f08a9-90c1-4b43-b933-dc4a77f3eb09.png">

The titles only get filled in once I visit each actor, which I have to do individually:

<img width="551" alt="Screenshot 2022-02-28 at 13 58 05" src="https://user-images.githubusercontent.com/647669/155987328-d06c7d50-b4e8-4444-9c21-656c9a006df7.png">

Not necessarily a blocker but wondering if there's a way we can improve on this? Just thinking about people who might have hundreds of entries that suddenly appear with non-functional titles after this update (in the second screenshot above, how would you know which entry to click to edit Meat Loaf, for example?).

mindctrl committed:
test: adjust test to check that title fields return post title instead of post meta value.

mindctrl committed:
refactor: move field value CRUD functions to field-functions.php.

mindctrl committed:
refactor: use helper function when fetching field values.

mindctrl committed:
refactor: use helper functions when fetching and saving field values.

mindctrl commented on includes/publisher/lib/field-functions.php:
```diff
@@ -222,3 +225,205 @@ function sanitize_field( string $type, $value ) {
+ *
+ * @return array|int|mixed|string
+ */
+function get_field_value( array $field, $post ) {
```
Not sure this is the best place for these functions, but I added them here for now. My original thought was put them in the global namespace and prefix them with `acm_` to make them easier to use. Anyone have thoughts? @wpengine/mintaka 

The reasons for extracting these out of the publisher class was: 1) these are needed outside the publisher class 2) the publisher class is a monster.

nickcernis replied:
I like the idea to extract these — thanks!

Not sure we need to expose them at the top level prefixed by `acm_` yet. I think we should continue to use namespaced functions until we know what we want to expose in a global API.

nickcernis commented:
@mindctrl New updates are working well for me on the publisher screen!

<img width="296" alt="Screenshot 2022-03-03 at 17 17 30" src="https://user-images.githubusercontent.com/647669/156610976-b905b545-dd61-49f3-b341-6ed0a4ff75fc.png">

Just had a question above about whether or not we should delete the title field meta. I think there may be value in preserving it so that it's accessible via the post meta field or post title, even if that means some duplication of data. Or we could filter the title field meta value to return the post title.

mindctrl committed:
chore: add 'String' case to switch statement.

If a field's type has changed from the default internal type
to a type to be registered with WPGraphQL, running it through
this switch again results in the `default` case running where
it is not desirable.
e.g. a `text` field type is converted to `String` before
being registered with WPGraphQL.
See `map_html_field_type_to_graphql_field_type()` for more info
on field types.

mindctrl committed:
refactor: remove early return for style consistency.

mindctrl committed:
test: confirm title field value is also available under field slug name.

mindctrl committed:
test: confirm title field value is also available under field slug name in REST responses.

mindctrl commented on includes/publisher/lib/field-functions.php:
```diff
@@ -222,3 +225,205 @@ function sanitize_field( string $type, $value ) {
+			}
+
+			if ( ! empty( $post->post_title ) || $post->post_title === $meta_value ) {
+				delete_post_meta( $post->ID, $field['slug'] );
```
Thanks for catching this. This was a side effect of refactoring the get/save logic. Since some other logic changes `$field['type']` to a WPGraphQL-friendly type, fetching the data in the resolver was triggering the `default` case in the switch — because, e.g., 'String' wasn't a match. See e480179c41a559c074fe3f6662bfe755d76e9ce0. I added a couple tests to check that title fields are also available under their field slug name.

nickcernis reviewed (dismissed):
Looking good!

<img width="295" alt="Screenshot 2022-03-14 at 13 44 36" src="https://user-images.githubusercontent.com/647669/158174829-f52f3328-5934-41f2-8333-39df9d1f9541.png">
<img width="431" alt="Screenshot 2022-03-14 at 13 44 45" src="https://user-images.githubusercontent.com/647669/158174836-ed7eca46-d741-4ea7-affe-c5dbd2b67a34.png">

```
> wp post get 1509
Cannot load Zend OPcache - it was already loaded
+-----------------------+------------------------------------------------+
| Field                 | Value                                          |
+-----------------------+------------------------------------------------+
| ID                    | 1509                                           |
| post_author           | 1                                              |
| post_date             | 2022-03-03 16:17:07                            |
| post_date_gmt         | 2022-03-03 16:17:07                            |
| post_content          |                                                |
| post_title            | New Rabbit                                     |
```

anthonyburchell reviewed (dismissed):
Looks great! Working well on my end. Big improvement!

mindctrl committed:
refactor: fix call to `delete_field_value()`

mindctrl committed:
refactor: revert removal of `$update` check in `set_post_attributes()`

mindctrl committed:
fix: restore auto-generated `entry{$post->ID}` post title.

mindctrl committed:
refactor: check for `entry{$post->ID}` title values before migrating title data.

mindctrl committed:
chore: return empty title value if post status is auto-draft.

This prevents Auto Draft from being displayed as the title
when Add New is clicked.

mindctrl committed:
chore: revert `set_post_attributes()` changes

nickcernis commented on includes/publisher/class-publisher-form-editing-experience.php:
```diff
@@ -602,48 +520,6 @@ public function render_feedback_notice(): void {
-
-		$fields = $this->models[ $post_type ]['fields'] ?? [];
-
-		$title_field = get_entry_title_field( $fields );
```
Is the `get_entry_title_field` function in use now with this code removal?

I don't see it anywhere except for the tests. Maybe we should remove/deprecate it if so.

luketimothyjones commented:
Is this issue now being addressed in another branch/PR?

mindctrl commented:
@luketimothyjones yes we're breaking it down into multiple PRs. Part of that was merged in a feature branch here https://github.com/wpengine/atlas-content-modeler/pull/479

Another PR is coming shortly to address some other parts.
